src/com/amazon/corretto/crypto/provider/AesCbcSpi.java

```diff
@@ -0,0 +1,685 @@
+      int outputOffset);
+
+  private static native int nUpdate(
+      long ctxPtr,
+      ByteBuffer inputDirect,
+      byte[] inputArray,
+      int inputOffset,
+      int inputLen,
+      int unprocessedInput,
+      ByteBuffer outputDirect,
+      byte[] outputArray,
+      int outputOffset);
+
+  @Override
+  protected byte[] engineDoFinal(byte[] input, int inputOffset, int inputLen)
+      throws IllegalBlockSizeException, BadPaddingException {
+    Utils.checkArrayLimits(emptyIfNull(input), inputOffset, inputLen);
+    finalOrUpdateStateCheck();
+    final byte[] result = new byte[getOutputSizeFinal(inputLen)];
+    final int resultLen = doFinal(null, emptyIfNull(input), inputOffset, inputLen, null, result, 0);
+    return resultLen == result.length ? result : Arrays.copyOf(result, resultLen);
+  }
+
+  @Override
+  protected int engineDoFinal(
+      byte[] input, int inputOffset, int inputLen, byte[] output, int outputOffset)
+      throws ShortBufferException, IllegalBlockSizeException, BadPaddingException {
+    Utils.checkArrayLimits(emptyIfNull(input), inputOffset, inputLen);
+    Utils.checkArrayLimits(output, outputOffset, output.length - outputOffset);
+    finalChecks(inputLen, output.length - outputOffset);
+
+    return doFinal(null, emptyIfNull(input), inputOffset, inputLen, null, output, outputOffset);
+  }
+
+  @Override
+  protected int engineDoFinal(ByteBuffer input, ByteBuffer output)
+      throws ShortBufferException, IllegalBlockSizeException, BadPaddingException {
+    finalChecks(input.remaining(), output.remaining());
+
+    final ShimByteBuffer inputShimByteBuffer = new ShimByteBuffer(input, true);
+    final ShimByteBuffer outputShimByteBuffer = new ShimByteBuffer(output, false);
+
+    final int result =
+        doFinal(
+            inputShimByteBuffer.directByteBuffer,
+            inputShimByteBuffer.array,
+            inputShimByteBuffer.offset,
+            input.remaining(),
+            outputShimByteBuffer.directByteBuffer,
+            outputShimByteBuffer.array,
+            outputShimByteBuffer.offset);
+
+    outputShimByteBuffer.writeBack(result);
+
+    input.position(input.limit());
+    output.position(output.position() + result);
+
+    return result;
+  }
+
+  private void finalChecks(final int inputLen, final int outputLen)
+      throws IllegalBlockSizeException, ShortBufferException {
+    finalOrUpdateStateCheck();
+    if (outputLen < getOutputSizeFinal(inputLen)) {
+      throw new ShortBufferException(outputLen + "<" + getOutputSizeFinal(inputLen));
+    }
+  }
+
+  private int getOutputSizeFinal(final int inputLen) throws IllegalBlockSizeException {
+    final long all = ((long) inputLen) + ((long) unprocessedInput);
+    final long rem = all % BLOCK_SIZE_IN_BYTES;
+    // If there is no padding or if we are decrypting, all the data must be aligned with block size.
+    if ((opMode == DEC_MODE || padding == NO_PADDING) && rem != 0) {
+      throw new IllegalBlockSizeException();
```

Comment:
Should we add an actionable message here for the users? Just informing them that padding is disabled and that doFinal was invoked with n-bytes remaining instead of the block size.

This is also a bit of an unexpected place for the boundary check. Suggest moving it somewhere more apparent.